core/services/workflows/syncer/workflow_registry.go

```diff
@@ -407,6 +398,11 @@ func (w *workflowRegistry) readRegistryEventsLoop(ctx context.Context, eventType
 			}
 
 			for _, event := range events {
+				isClosed := ctx.Err()
```

Comment:
You can just do the same select dance you did on receipt:

```
select {
case <-ctx.Done():
  // handle done
case w.eventCh <- event.Event:
}
```

The code you've written has a race condition between the two calls which means you could still hang if ctx is cancelled shortly after you've checked it.